### PR TITLE
Optimization of Opensles for old Android phones, and some Chinese AOSP

### PR DESCRIPTION
**Pull request by KORuL**

### CODE REVIEW

**Comment by sreimers** [comment]
There are many linter warnings, you can reproduce locally with:
```
$ test/ccheck.py modules
./modules/opensles/recorder.c:89: C++ comment, use C comments /* ... */ instead
./modules/opensles/recorder.c:90: C++ comment, use C comments /* ... */ instead
./modules/opensles/recorder.c:91: C++ comment, use C comments /* ... */ instead
./modules/opensles/recorder.c:99: line is too wide (127 - max 79)
./modules/opensles/recorder.c:100: line is too wide (max 79)
./modules/opensles/recorder.c:101: line is too wide (100 - max 79)
./modules/opensles/recorder.c:102: line is too wide (97 - max 79)
./modules/opensles/recorder.c:103: line is too wide (136 - max 79)
./modules/opensles/recorder.c:117: line is too wide (103 - max 79)
./modules/opensles/recorder.c:118: line is too wide (91 - max 79)
./modules/opensles/recorder.c:120: line is too wide (86 - max 79)
./modules/opensles/recorder.c:123: line is too wide (147 - max 79)
./modules/opensles/recorder.c:127: line is too wide (148 - max 79)
./modules/opensles/recorder.c:133: line is too wide (114 - max 79)
./modules/opensles/recorder.c:134: line is too wide (88 - max 79)
./modules/opensles/recorder.c:142: line is too wide (110 - max 79)
./modules/opensles/recorder.c:143: line is too wide (89 - max 79)
...
```

File: modules/opensles/recorder.c [hunk 1]
```diff
@@ -107,6 +112,49 @@ static int createAudioRecorder(struct ausrc_st *st, struct ausrc_prm *prm)
...
+	}
+
+	SLAndroidNoiseSuppressionItf player_noise;
+	r = (*st->recObject)->GetInterface(st->recObject,
+									   SL_IID_ANDROIDAUTOMATICGAINCONTROL,
+									   &player_noise);
+	if (SL_RESULT_SUCCESS == r) {
+		SLboolean enable = SL_BOOLEAN_TRUE;
+		(*player_noise)->SetEnabled(player_noise, enable);
+	}
```
Comment (alfredh) [inline modules/opensles/recorder.c:156]: AEC, AEG and NS should be configurable, either via baresip core or only for this module.

For example running two instances of AEC (this one and webrtc_aec) can lead
to bad audio and/or interference.

Comment (juha-h) [reply modules/opensles/recorder.c:156]: Yes, they should be configurable.  For example, here is how AEC is set in Google test code:
```
    /* Enable AEC if requested */
    if (aec) {
        SLAndroidAcousticEchoCancellationItf aecItf;
        result = (*recorder)->GetInterface(
                recorder, SL_IID_ANDROIDACOUSTICECHOCANCELLATION, (void*)&aecItf);
        printf("AEC is %savailable\n", SL_RESULT_SUCCESS == result ? "" : "not ");
        if (SL_RESULT_SUCCESS == result) {
            result = (*aecItf)->SetEnabled(aecItf, true);
            ExitOnError(result);
            SLboolean enabled;
            result = (*aecItf)->IsEnabled(aecItf, &enabled);
            ExitOnError(result);
            printf("AEC is %s\n", enabled ? "enabled" : "not enabled");
        }
    }
```
Same kind if debug messages would be useful also barersip module.

**Comment by juha-h** [comment]
Why is none of these set?
```
/*---------------------------------------------------------------------------*/
/* Android AudioRecorder configuration                                       */
/*---------------------------------------------------------------------------*/

/** Audio recording preset */
/** Audio recording preset key */
#define SL_ANDROID_KEY_RECORDING_PRESET ((const SLchar*) "androidRecordingPreset")
/** Audio recording preset values */
/**   preset "none" cannot be set, it is used to indicate the current settings
 *     do not match any of the presets. */
#define SL_ANDROID_RECORDING_PRESET_NONE                ((SLuint32) 0x00000000)
/**   generic recording configuration on the platform */
#define SL_ANDROID_RECORDING_PRESET_GENERIC             ((SLuint32) 0x00000001)
/**   uses the microphone audio source with the same orientation as the camera
 *     if available, the main device microphone otherwise */
#define SL_ANDROID_RECORDING_PRESET_CAMCORDER           ((SLuint32) 0x00000002)
/**   uses the main microphone tuned for voice recognition */
#define SL_ANDROID_RECORDING_PRESET_VOICE_RECOGNITION   ((SLuint32) 0x00000003)
/**   uses the main microphone tuned for audio communications */
#define SL_ANDROID_RECORDING_PRESET_VOICE_COMMUNICATION ((SLuint32) 0x00000004)
/**   uses the main microphone unprocessed */
#define SL_ANDROID_RECORDING_PRESET_UNPROCESSED         ((SLuint32) 0x00000005)
```
Second to last might be the right one.

File: modules/opensles/recorder.c (hunk 1 above)
Comment (juha-h) [reply modules/opensles/recorder.c:156]: I added info message here:
```
	SLAndroidAcousticEchoCancellationItf player_aec;
	r = (*st->recObject)->
		GetInterface(st->recObject,
			     SL_IID_ANDROIDACOUSTICECHOCANCELLATION,
			     &player_aec);
	info("AEC is %savailable\n", SL_RESULT_SUCCESS == r ? "" : "not ");
```
And on my Android 10 device got:
```
10-02 09:59:01.086  5275  5321 D Baresip Lib: opensles: opening recorder 16000Hz, 1channels
10-02 09:59:01.086  5275  5321 D Baresip Lib: AEC is not available
```
So are these features only available on old Android devices?

Comment (KORuL) [reply modules/opensles/recorder.c:156]: Yes, on my Chinese tablet log - is enabled true

**Comment by KORuL** [comment]
> Why is none of these set?
> 
> ```
> /*---------------------------------------------------------------------------*/
> /* Android AudioRecorder configuration                                       */
> /*---------------------------------------------------------------------------*/
> 
> /** Audio recording preset */
> /** Audio recording preset key */
> #define SL_ANDROID_KEY_RECORDING_PRESET ((const SLchar*) "androidRecordingPreset")
> /** Audio recording preset values */
> /**   preset "none" cannot be set, it is used to indicate the current settings
>  *     do not match any of the presets. */
> #define SL_ANDROID_RECORDING_PRESET_NONE                ((SLuint32) 0x00000000)
> /**   generic recording configuration on the platform */
> #define SL_ANDROID_RECORDING_PRESET_GENERIC             ((SLuint32) 0x00000001)
> /**   uses the microphone audio source with the same orientation as the camera
>  *     if available, the main device microphone otherwise */
> #define SL_ANDROID_RECORDING_PRESET_CAMCORDER           ((SLuint32) 0x00000002)
> /**   uses the main microphone tuned for voice recognition */
> #define SL_ANDROID_RECORDING_PRESET_VOICE_RECOGNITION   ((SLuint32) 0x00000003)
> /**   uses the main microphone tuned for audio communications */
> #define SL_ANDROID_RECORDING_PRESET_VOICE_COMMUNICATION ((SLuint32) 0x00000004)
> /**   uses the main microphone unprocessed */
> #define SL_ANDROID_RECORDING_PRESET_UNPROCESSED         ((SLuint32) 0x00000005)
> ```
> 
> Second to last might be the right one.



> Why is none of these set?
> 
> ```
> /*---------------------------------------------------------------------------*/
> /* Android AudioRecorder configuration                                       */
> /*---------------------------------------------------------------------------*/
> 
> /** Audio recording preset */
> /** Audio recording preset key */
> #define SL_ANDROID_KEY_RECORDING_PRESET ((const SLchar*) "androidRecordingPreset")
> /** Audio recording preset values */
> /**   preset "none" cannot be set, it is used to indicate the current settings
>  *     do not match any of the presets. */
> #define SL_ANDROID_RECORDING_PRESET_NONE                ((SLuint32) 0x00000000)
> /**   generic recording configuration on the platform */
> #define SL_ANDROID_RECORDING_PRESET_GENERIC             ((SLuint32) 0x00000001)
> /**   uses the microphone audio source with the same orientation as the camera
>  *     if available, the main device microphone otherwise */
> #define SL_ANDROID_RECORDING_PRESET_CAMCORDER           ((SLuint32) 0x00000002)
> /**   uses the main microphone tuned for voice recognition */
> #define SL_ANDROID_RECORDING_PRESET_VOICE_RECOGNITION   ((SLuint32) 0x00000003)
> /**   uses the main microphone tuned for audio communications */
> #define SL_ANDROID_RECORDING_PRESET_VOICE_COMMUNICATION ((SLuint32) 0x00000004)
> /**   uses the main microphone unprocessed */
> #define SL_ANDROID_RECORDING_PRESET_UNPROCESSED         ((SLuint32) 0x00000005)
> ```
> 
> Second to last might be the right one.

image

**Comment by juha-h** [comment]
KORuL writes:

> image

That is about performance, not recording.

**Comment by juha-h** [comment]
KORuL writes:

> Yes, on my Chinese tablet log - is enabled true

I tried all three and they were not available on my Android 10 device has
pure Android w/o any vendor specific stuff.

So clearly webrtc_aecm module cannot be replaced by the OpenSLES
setting.

**Comment by alfredh** [comment]
please make a clean PR with small changes.

The changes should be documented, i.e. "this change improves X on Tablet Y".

**Pull request closed without merging**